platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/operations/TestListenerBuildOperationAdapter.java

```diff
@@ -49,34 +48,37 @@ public class TestListenerBuildOperationAdapter implements TestListenerInternal {
     private final Map<TestDescriptor, InProgressExecuteTestBuildOperation> runningTests = new HashMap<TestDescriptor, InProgressExecuteTestBuildOperation>();
     private final BuildOperationListener listener;
     private final BuildOperationIdFactory buildOperationIdFactory;
-    private final Clock clock;
 
-    public TestListenerBuildOperationAdapter(BuildOperationListener listener, BuildOperationIdFactory buildOperationIdFactory, Clock clock) {
+    public TestListenerBuildOperationAdapter(BuildOperationListener listener, BuildOperationIdFactory buildOperationIdFactory) {
         this.listener = listener;
         this.buildOperationIdFactory = buildOperationIdFactory;
-        this.clock = clock;
     }
 
     @Override
     public void started(final TestDescriptorInternal testDescriptor, TestStartEvent startEvent) {
-        long currentTime = clock.getCurrentTime();
         BuildOperationDescriptor testBuildOperationDescriptor = createTestBuildOperationDescriptor(testDescriptor, startEvent);
-        runningTests.put(testDescriptor, new InProgressExecuteTestBuildOperation(testBuildOperationDescriptor, currentTime));
-        listener.started(testBuildOperationDescriptor, new OperationStartEvent(currentTime));
+        runningTests.put(testDescriptor, new InProgressExecuteTestBuildOperation(testBuildOperationDescriptor, startEvent.getStartTime()));
+        listener.started(testBuildOperationDescriptor, new OperationStartEvent(startEvent.getStartTime()));
     }
 
     @Override
     public void completed(TestDescriptorInternal testDescriptor, TestResult testResult, TestCompleteEvent completeEvent) {
-        long currentTime = clock.getCurrentTime();
         InProgressExecuteTestBuildOperation runningOp = runningTests.remove(testDescriptor);
-        listener.finished(runningOp.descriptor, new OperationFinishEvent(runningOp.startTime, currentTime, testResult.getException(), new Result(testResult)));
+        // Adjusted to not use current time, as that may be wrong in the context of remote test execution. The time given by the test result is more reliable.
+        long endTime = completeEvent.getEndTime();
+        if (runningOp.startTime > endTime) {
+            endTime = runningOp.startTime;
+        }
+        listener.finished(runningOp.descriptor, new OperationFinishEvent(runningOp.startTime, endTime, testResult.getException(), new Result(testResult)));
     }
 
     @Override
     public void output(final TestDescriptorInternal testDescriptor, final TestOutputEvent event) {
-        long currentTime = clock.getCurrentTime();
         InProgressExecuteTestBuildOperation runningOp = runningTests.get(testDescriptor);
-        listener.progress(runningOp.descriptor.getId(), new OperationProgressEvent(currentTime, new OutputProgress(event)));
+        if (runningOp == null) {
+            throw new IllegalStateException("Received output for test that is not running: " + testDescriptor);
```

Comment:
This is a helpful debug statement for callers that accidentally pass the wrong ID to `StateTrackingTestResultProcessor` (I was passing the `Descriptor` instead of the generated ID object). Originally, because `StateTrackingTestResultProcessor` returns a `new UnknownTestDescriptor()` in this case, it would crash with an NPE.

Perhaps we should just adjust `StateTrackingTestResultProcessor`, but I wanted to avoid messing with it too much.